### PR TITLE
[WIP] Adapt SEraster for Seurat spatial objects

### PR DESCRIPTION
**Pull request by zskylarli**

### CODE REVIEW

File: R/SEraster.R [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+      image <- ifelse(is.null(image), Images(spe, assay=assay_name)[1], image)
+      coords <- GetTissueCoordinates(spe, image = image)
+      if("cell" %in% colnames(coords)){
+        rownames(coords) <- coords$cell
+      }
+      
+      if (is.null(assay_name)) {
+        assay_name <- DefaultAssay(spe)
+        counts <- LayerData(spe[[assay_name]], layer = slot)
+        out <- SEraster::rasterizeMatrix(counts, coords, bbox = bbox_common, resolution = resolution, square = square, fun = fun, n_threads = n_threads, BPPARAM = BPPARAM, verbose = verbose)
```
Comment (dcollins15) [inline R/SEraster.R:122]: nit: line length

```suggestion
        out <- SEraster::rasterizeMatrix(
          counts, 
          coords, 
          bbox = bbox_common, 
          resolution = resolution, 
          square = square, 
          fun = fun, 
          n_threads = n_threads, 
          BPPARAM = BPPARAM, 
          verbose = verbose
        )
```

**Review by dcollins15 — COMMENTED** [review]
Looking good so far! 🙌 

Since this is still a WIP I'll just comment on the `SpatialImage`/`FOV` typing stuff - also happy to chat about eliminating code duplication if that would be helpful 🙂 

tl;dr: Make sure to pass `scale = NULL` when calling `GetTissueCoordinates` and `Radius` and consider just outputting `FOV` and `VisiumV2` instances (i.e update `VisiumV1` inputs on the way through)

Just to recap the state of things: `Seurat`/`SeuratObject` define two interfaces (by which I mean structural types) that inherit the `SpatialImage` virtual class. All concrete `SpatialImage` classes implement a common set of generic functions - we'll focus on `GetTissueCoordinates`, `Radius`, and `GetImage`:

1. The original spatial interface is implemented by `SlideSeq`, `STARmap`, `VisiumV1`, and `VisiumV2`
- `GetTissueCoordinates` returns a `data.frame` whose first column contains X coordinates, second columns contains Y coordinates, and `rownames` contains unique barcodes from the corresponding assay
- `GetTissueCoordinates` accepts a `scale` param, or else ignores it via `...` (this was the case pre `Seurat` v5.1)
- `Radius` accepts a `scale` param, or else ignores it via `...` (added to accommodate hires images from Visium)
- `GetImage` returns list of (potentially empty) image values
- Relied on by `SpatialPlot`
- **I would strongly recommend against coupling this wrapper to this interface** - 
2. `FOV` which supports various image-based spatial transcriptomic technologies - also implemented by `VisiumV2`
- `GetTissueCoordinates` returns a `data.frame` with ordered columns `"x"`, `"y"`, and `"cell"` - `rownames` may or may not be set to the `"cell"` column which contains unique barcodes from the corresponding assay
- `Radius` accepts a `scale` param, or else is not defined - `Radius.Centroids` ignores all params via `...`
- `GetImage` returns a list of image values, or else is not defined

Eventually, we'd like to deprecate the first interface in favor of the second but only after updating the latter so that:
- `GetTissueCoordinates` and `Radius` can both be passed `scale = {TRUE/FALSE}` 
- `GetImage` returns list of (potentially empty) image values

All of that is to say, the main change you'll want to make is to pass `scale = NULL` into `GetTissueCoordinates` and `Radius` anytime you're accessing either property so that you don't have to muck around with any scaling logic inside `createRasterizedObjects`

I'd also suggest that you avoid trying to output `VisiumV1` instances and then update `createRasterizedObjects` so that it builds an `FOV` by default and then updates it into a `VisiumV2` instance in a `tryCatch` block alongside a call to `ScaleFactors` which is only defined for `VisiumV1`/`VisiumV2` images. 

In effect, `VisiumV1` inputs would get rasterized into `VisiumV2` ouputs and everything else would become a `FOV`. In the immediate term this would be bad because you can't give a `FOV` to `SpatialPlot` but that's only because the output of `GetTissueCoordinates.FOV` doesn't have `rownames` set which should be a straight-forward patch to make in `SeuratObject` (I was trying to make as few changes as possible for v5.0.2 but I kinda regret not including this one). 

On a semi-related note we might want to consider limiting the output to just `FOV` - if we're also manipulating the underlying image values then this won't be practical but I'm not totally sure what the story with those is meant to be exaclty 🤷 It's easy enough to convert a `FOV` to a `VisiumV2` given the missing info:

```R
fov_to_visium <- function(fov, image, scale.factors) {
  # build the final `VisiumV2` - essentially just adding `image` and
  # `scale.factors` to the object
  visium.fov <- new(
    Class = "VisiumV2",
    boundaries = fov@boundaries,
    molecules = fov@molecules,
    assay = fov@assay,
    key = fov@key,
    image = image,
    scale.factors = scale.factors
  )
}
```

File: R/SEraster.R [hunk 2]
```diff
@@ -0,0 +1,492 @@
+#' @include internal.R
+#'
+NULL
+#' createRasterizedObject
+#' @keyword internal
+#' 
+createRasterizedObject <- function(input, out, name) {
```
[the remark this answers is not in the document]
Comment (zskylarli) [reply R/SEraster.R:7]: Yep, thanks for reminding! - and `name` should probably be `new_assay_name` and/or be optional since I just realized `rasterizeCellType` passes in the metadata column name but it's never used (the condition under it is just to pull the right image to associate the assay with).

File: R/SEraster.R [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+      type = c("segmentation", "centroids"),
+      radius = radius,
+      molecules = input[[image]]$molecules,
+      assay = name,
+      key = Key("rasterized_", quiet = TRUE)
+    )
+  }
+  output@images[[paste0(name,".rasterized")]]<- new.fov
+  return(output)
+}
```
[the remark this answers is not in the document]
Comment (zskylarli) [reply R/SEraster.R:84]: The reason I didn't do this was because existing VisiumV1 class objects (like stxbrain in SeuratData) don't have Centroids, and VisiumV2 objects don't have segmentations because that's only relevant to in-situ data, so you wouldn't be able to extract info with 

```
input_fov <- input[[image_name]]
input_centroids <- input.fov[["centroids"]]
input_segmentation <- input.fov[["segmentation"]]
input_molecules <- input.fov[["molecules"]]
```

Not sure if there is really an alternative but if you have other ideas please let me know!

[the remark this answers is not in the document]
Comment (dcollins15) [reply R/SEraster.R:84]: It should be pretty easy to essentially ignore the "segmentation"s and pass them straight through:

```R
boundary_names <- Boundaries(input_fov)
input_boundaries <- input_fov@boundaries
input_centroids <- ifelse(
    "centroids" %in% boundary_names,
    input_boundaries[["centroids"]],
    NULL
)
input_molecules <- input_fov@molecules
```

Wrapping the whole thing in a `tryCatch` allows us to accomodate non-`FOV` images:

```R
tryCatch(
    {
      boundary_names <- Boundaries(input_fov)
      input_boundaries <- input_fov@boundaries
      input_centroids <- ifelse(
        "centroids" %in% boundary_names,
        input_boundaries[["centroids"]],
        NULL
      )
      input_molecules <- input_fov@molecules
    },
    error = function(e) {
      input_boundaries <- list()
      input_centroids <- NULL
      input_molecules <- NULL
    }
  )
```

without having to know the actual type being handed around:

```R
if (!is.null(input_centroids)) {
    output_nsides <- input_centroids@nsides
    output_theta <- input_centroids@theta

    output_radius <- sqrt(
      nrow(input[[]]) / nrow(meta_rast)
    ) * Radius(input_centroids) # unscaled
  } else {
    nsoutput_nsidesides <- NULL
    output_theta <- NULL

    output_radius <- Radius(input_fov)
  }

  output_centroids <- CreateCentroids(
    coords = output_coordinates,
    nsides = output_nsides,
    radius = output_radius,
    theta = output_theta
  )
```

I guess you could technically also get an `FOV` with _multiple_ `Molecules` but all that means is we should avoid using `CreateFOV` and just call `new` directly (also note the way "segmentations" will be handed through if it's present):

```R
output_boundaries <- input_boundaries
  output_boundaries[["centroids"]] <- output_centroids

  output_fov <- new(
    Class = "FOV",
    boundaries = output_boundaries,
    molecules = input_molecules,
    assay = assay,
    key = Key(output_image_name, quiet = TRUE)
  )
```

**Comment by dcollins15** [comment]
Another edge case to keep in mind is that not all `FOV` instances will have `centroids` and (correct me if I'm wrong) `SEraster` doesn't handle cell segmentations. Whenever we make a call to `GetTissueCoordinates.FOV`/`GetTissueCoordinates.VisiumV2` we need to be specifying `which = "centroids"` (and check that this doesn't raise any errors). 

What is the expected behavior in this case?

**Comment by zskylarli** [comment]
> Another edge case to keep in mind is that not all `FOV` instances will have `centroids` and (correct me if I'm wrong) `SEraster` doesn't handle cell segmentations. Whenever we make a call to `GetTissueCoordinates.FOV`/`GetTissueCoordinates.VisiumV2` we need to be specifying `which = "centroids"` (and check that this doesn't raise any errors).
> 
> What is the expected behavior in this case?

Will check with the main author but seems counterintuitive to rasterize segmentations.

Also, have tried integrating your fix, but it ends up being that there are too many differences between VisiumV2 instance vs a normal FOV that it is almost cleaner to just implement by class than it is to use a lot of different `tryCatch`s; for example, the lack of a segmentation for V2 becomes an issue with `CreateFOV`.

File: R/SEraster.R [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+  if (is.list(input)) {
+    ## create a common bbox
+    bbox_mat <- do.call(rbind, lapply(seq_along(input), function(i) {
+      pos <- GetTissueCoordinates(input[[i]])
+      if (!is.null(names(input))) {
+        dataset <- names(input)[[i]]
+      } else {
+        dataset <- i
+      }
+      return(data.frame(dataset = dataset, xmin = min(pos[,1]), xmax = max(pos[,1]), ymin = min(pos[,2]), ymax = max(pos[,2])))
```
[the remark this answers is not in the document]
Comment (zskylarli) [reply R/SEraster.R:99]: This is from the original code so I'm not going to change it

File: DESCRIPTION [hunk 5]
```diff
@@ -1,14 +1,15 @@
...
+Version: 0.3.6
+Date: 2024-05-10
 Authors@R: c(
   person(given = 'Andrew', family = 'Butler', email = 'abutler@nygenome.org', role = 'aut', comment = c(ORCID = '0000-0003-3608-0463')),
   person(given = "Saket", family = "Choudhary", email = "schoudhary@nygenome.org", role = "ctb", comment = c(ORCID = "0000-0001-5202-7633")),
   person(given = 'David', family = 'Collins', email = 'dcollins@nygenome.org', role = 'ctb', comment = c(ORCID = '0000-0001-9243-7821')),
   person(given = "Yuhan", family = "Hao", email = "yhao@nygenome.org", role = "ctb", comment = c(ORCID = "0000-0002-1810-0822")),
   person(given = "Austin", family = "Hartman", email = "ahartman@nygenome.org", role = "ctb", comment = c(ORCID = "0000-0001-7278-1852")),
   person(given = 'Paul', family = 'Hoffman', email = 'nygcSatijalab@nygenome.org', role = c('aut', 'cre'), comment = c(ORCID = '0000-0002-7693-8957')),
+  person(given = 'Skylar', family = 'Li', email = 'sli@nygenome.org', role = 'ctb', comment = c(ORCID = '0009-0002-3486-5785'))
```
Comment (dcollins15) [inline DESCRIPTION:12]:
```suggestion
  person(given = 'Skylar', family = 'Li', email = 'sli@nygenome.org', role = 'ctb', comment = c(ORCID = '0009-0002-3486-5785')),
```

**Review by dcollins15 — CHANGES_REQUESTED** [review]
> Also, have tried integrating your fix, but it ends up being that there are too many differences between VisiumV2 instance vs a normal FOV that it is almost cleaner to just implement by class than it is to use a lot of different tryCatchs; for example, the lack of a segmentation for V2 becomes an issue with CreateFOV.

If we were able to use `CreateFOV` why wouldn't we just always hand in a named list to the `coords` param? I don't think we're able to use the function though - see my comment about the handling of `molecules`. 

You also _are_ using the `tryCatch` to handle `VisiumV2` vs. `FOV`? In terms of the `VisiumV1`-specific logic, I am strongly against this. Why would we support `VisiumV1` instances but not `SlideSeq` or `STARmap`? 

> Will check with the main author but seems counterintuitive to rasterize segmentations.

If this is the case, then shouldn't we be dropping the segmentations? Since new cell barcodes are being generated, the segmentations won't match and you eventually get an error if you try to set the `FOV`'s `DefaultBoundary` to "segmentation":

```
Error in `[[<-`:
! Cannot add new cells with [[<-
```

Similarly if `DefaultBoundary` is set to "segmentation" for the input object, then you'll also get an error:

```
Error in `.rowNamesDF<-`(x, value = value) : 
  duplicate 'row.names' are not allowed
In addition: Warning message:
non-unique values when setting 'row.names': ‘10022’, ‘10023’ [... truncated] 
```

**Comment by zskylarli** [comment]
Okay, I think I understand what you mean now and have tried to implement the suggestions. One issue remaining is that I can't figure out how the radius is used for plotting FOV objects - I see that the `Radius()` function is called within `GeomSpatial` to get the `spot.size` for SpatialPlots, but I actually have no idea how this is handled for SingleImagePlots; assuming it's different since it's inheriting ggplot's `geom_point`? Let me know if you have any ideas.

**Comment by dcollins15** [comment]
> Okay, I think I understand what you mean now and have tried to implement the suggestions.

At a quick glance `createRasterizedObject` is looking pretty slick now - nicely done. 

> One issue remaining is that I can't figure out how the radius is used for plotting FOV objects - I see that the Radius() function is called within GeomSpatial to get the spot.size for SpatialPlots, but I actually have no idea how this is handled for SingleImagePlots; assuming it's different since it's inheriting ggplot's geom_point instead of using our internal geom_spatial? Let me know if you have any ideas.

I think it may just be ignored? `ImageFeaturePlot` and `ImageDimPlot` both have a `size` parameter that appears to get handed down to `geom_point` inside `SingleImagePlot`. Are you running into plotting issues post-rasterization?

**Comment by zskylarli** [comment]
Yeah, just that the radius isn't really passed into the plotting functions, so adjusting size will scale the points correctly but adjusting the radius for spots in the rasterized object doesn't actually seem to do anything.

**Comment by dcollins15** [comment]
noice

The vignette is easy to follow and everything seems to be working as expected 👌 I'm still in the process of writing out some suggestions to tidy the code up some but I've got a couple of high-level comments that are worth considering in the meantime:

1. Namespace collisions - we probably don't want to _have_ to make calls to `SeuratWrappers::` in the vignette. We wouldn't have to if we didn't import the `SERaster` package or otherwise updated the `SeuratWrapper` methods to use PascalCase instead of camelCase. Since `Seurat` usually uses PascalCase for function names this might be kind of an elegant way to go. 

2. `permutateByRotation` is currently clobbering the image cell names with simple indices. This means that until `rasterizeGeneExpression` or `rasterizeCellType` is called, the `Seurat` instances are invalid and cannot be passed to any function that checks `` (e.g. `NormalizeData`). Furthermore, the fact that the clobbered cell names still work with the downstream methods might suggest that `rasterizeGeneExpression` and `rasterizeCellType` will not work as expected if handed an object where the counts index differs from the coordinates index (e.g. the image has been previously cropped).

3. I think it might be more ergonomic to have `permutateByRotation` return a single `Seurat` instance but with multiple `FOV`s. Then, instead of allowing a list to be passed to `rasterizeGeneExpression` and `rasterizeCellType` we just have to support multiple image names to be specified. This isn't possible to do with a `SpatialExperiment` so this would be a small departure from the underlying API, but I think this is a better way to capture the workflow with our data model. 

Once I've got my final code suggestions pushed up we can touch base and settle on action items to try to get this merged in as soon as practical 👍

**Review by dcollins15 — COMMENTED** [review]
I took a closer read through the underlying `SEraster` package and noticed some concerns regarding the way we’re essentially forking `rasterizeGeneExpression`, `rasterizeCellType`, and `permutateByRotation`. By implementing things this way, we are committing ourselves to manually keeping the two implementations in sync. For really simple changes (e.g., changing the default values of parameters), this is manageable but can be a bit tedious. For more complicated updates, resolving the differences between two slightly different versions of the same method can become very challenging.

On a positive note, your decision not to re-implement rasterizeMatrix was a good one! 👍

I understand you might not want to spend too much more time on this, so I won’t block you here. If I were you, I would consider either:

Decoupling the `SeuratWrapper` functions from the `SEraster` so that the wrapper constitutes a new interface (similar to how previous wrappers are implemented, see https://github.com/KlugerLab/ALRA/blob/master/R/alra.R)
Implementing `rasterizeGeneExpression` and `rasterizeCellType` as true wrappers that call the equivalent SEraster methods, which would involve converting to and from a `SpatialExperiment` instance.

Given the timing, I realize either option might not seem very attractive, but I think it’s still valuable to be mindful of any technical debt we knowingly incur. 

All that said, I think I’d be ok with merging this in once `permutateByRotation` has been updated to avoid clobbering cell barcodes - I think it’s just a function of making sure that the `rownames` are set for the `data.frame` returned by `GetTissueCoordinates` 👌

All in all, nice work!

**Review by dcollins15 — COMMENTED** [review]
Per our discussion last Friday, we'll rename the main methods to avoid conflicting with the underlying `SEraster` package:

- `rasterizeGeneExpression` -> `RunRasterizeGeneExpression`
- `rasterizeCellType` -> `RunRasterizeCellType`
- `permutateByRotation` -> `RunPermutateByRotation`

In addition, I've hammered out a small update that consolidates the `updateFOV` and `updateVisiumV2` 👍
